### ISSUE TITLE
The call to `localtime_r` may be unsound

### ISSUE THREAD

**Issue by quininer** [chronotope/chrono#499]
I found that `getenv` and `setenv` in libc are not thread-safe [1], and most impl of `localtime_r` in libc directly call `getenv` [2]. This means that `localtime_r` may have data race with `setenv`.

In order to ensure soundness of `setenv`, libstd add a lock to it [1], but this means that using `getenv` without libstd will be unsound. 

This problem is not easy to reproduce on glibc, because glibc's `localtime_r` caches timezone. but using musl can easily reproduce it.

1. https://github.com/rust-lang/rust/pull/24741
2. https://github.com/aosp-mirror/platform_bionic/blob/master/libc/tzcode/localtime.c#L1321 and https://git.musl-libc.org/cgit/musl/tree/src/time/__tz.c#n127

POC: https://gist.github.com/quininer/2063c31b0bc1753989122e782b182bea

**Comment by dekellum** [comment]
Rust libstd has a similar, long open issue, for its own calls to `getaddrinfo`: rust-lang/rust#27970

**Comment by quodlibetor** [comment]
Thank you for this!

From reading the associated issues, it seems like touching the environment is the thing that glibc actually considers to be non-thread-safe, is that right? In particular https://github.com/rust-lang/rust/issues/27970#issuecomment-134273584 in particular, it seems like libc wants modifying the environment to be illegal in a multithreaded environment:

> functions that modify the environment are all marked with const:env and regarded as unsafe. Being unsafe, the latter are not to be called when multiple threads are running or asynchronous signals are enabled, and so the environment can be considered effectively constant in these contexts, which makes the former safe.

I agree that the solution in the rust stdlib getaddrinfo issue (being able to take the env lock) would help.

I'm trying to think of a good solution here, and it's hard because none of the underlying libraries (libc or stdlib) provide us with the necessary tools. In particular, even if the stdlib provided us with access to the env lock, any other code that interacts with libc directly could inadvertently call `setenv` and leave us in the same situation. That would qualify as going outside of the safety properties rust is supposed to provide, though.

**Comment by Arnavion** [comment]
Actually, this conversation is founded on a wrong assumption. chrono has the issue not because of its time 0.1 dependency, but because [it directly calls `localtime_r`]() in the same way that time 0.1 does. AFAICT the code from time 0.1 that calls `localtime_r` is never called by chrono, at least not in the `Local::now` call path.

So unless I'm missing something, while it's unfortunate that the PR to update to time 0.3 hasn't been merged for a long time, it also won't actually fix the crash reported here.

**Comment by Arnavion** [comment]
For everyone landing here from the chrono security advisory RUSTSEC-2020-0159, note that the issue with chrono is not because of its dependency on time 0.1. It's because chrono itself calls `localtime_r` in a way that can have the issues described in the advisory. While time 0.1 also has an advisory (RUSTSEC-2020-0071) for calling `localtime_r`, and while chrono does default to enabling a dependency on time 0.1, chrono never calls the function in time 0.1 that would then call `localtime_r`. See https://github.com/chronotope/chrono/issues/602#issuecomment-940445390

That means that hoping for chrono to update to time 0.3 won't fix the issue. It also means that disabling the oldtime feature of chrono will not fix the issue.

To give some idea about how vulnerable your chrono-using program is, note that the issue happens specifically when two things happen simultaneously:

1. Something in your program changes the process environment. This could be via `std::env::set_var`, or directly with libc, or something that isn't even a Rust thread.

1. Something in your program program calls `chrono::Local::now()`

If your program does not have both those things happen simultaneously, it will not trigger the issue. (You will of course need to determine this by looking at not just your program's code but also all your deps' code.) Notably, `chrono::Local::now()` is the only code path in which chrono calls `localtime_r`. The rest of chrono's API does not call it.

**Comment by Zooce** [comment]
I'm just thinking out loud here - maybe it will trigger a thought in someone who knows more about this crate -- Since it sounds like a real fix is non-trivial and could take quite some time to address, would grouping the different offsets into features help to mitigate this in the meantime?

I'm only using UTC in my project, so the issue doesn't affect me, but we do use cargo-audit and we get flagged simply for using chrono at all. Not sure cargo-audit takes into account the features actually used in a Cargo.toml, so this may useless in that case as well, but it's just thought.

**Comment by Arnavion** [comment]
>There is also no point to separating getting the localtime out to a separate crate. Either you accept that calling localtime requires environment variables to be static (or for your program to be single-threaded) or you don't ever get the local time or figure out how far your computer is from GMT.

I'm not saying the crate should be split up either. But it's worth noting that it has the advantage that a program that only deals with `DateTime<Utc>` can be statically (as in via its Cargo.lock) sure that it doesn't have any call path to `Local::now()`. Tooling that only looks at the lockfile, like cargo-audit, would benefit in that case. Software that only deals solely with `DateTime<Utc>` is quite common.

>We could add more docs around this

Yes. Documenting the chrono function as "might call into libc API that is not safe to call if the environment is being changed simultaneously" is probably the best bet, just like `std::env::set_var` does.

>IMO this rustsec advisory is misleading, and for many people it's unactionable.

The advisory is unactionable in that there is no version of chrono that a user can update to. But it *is* actionable in that a user can check their code to be sure that the conditions for the issue (environ changing at the same time as `Local::now()` being called) do not occur.

I agree that it's misleading. It's vague, doesn't mention the conditions for the issue to occur, and links to an issue in the time crate's repo. Even now people are being confused and thinking the problem can be resolved by disabling the `oldtime` feature or updating the time dep to v0.3 ...

>It is technically broken in multithreaded environments that change env vars after spawning multiple threads (which IIRC is undefined for libc in general anyway) and it would be nice to fix, but (AFAIK) fixing it would rely on fixing or reimplementing libc.

>IMO the security advisory would be more accurate if it was for the standard library which attempts to allow setting env vars after fork and is the thing that is actually undefined wrt libc.

Multi-threaded C programs calling `setenv` is never going to change. It's up to the programmer to ensure that those calls do not race with anything that would be affected. There's already Rust code that both relies on changing the environ in a multi-threaded program (unit tests, C library interop, etc) and already makes sure to avoid this problem, so penalizing it by disabling `std::env::set_var` in the presence of threads won't really help. (The rust-lang/rust issue about `getaddrinfo` already discussed this idea.)

>Completely eliminating the functionality around local times and offsets is also unacceptible for Chrono, so there's nothing really that can be done here. Unless I'm missing something.

I agree. For someone that does want to use `DateTime<Local>`, they should not be prevented from calling `Local::now()` if they've taken the precautions to ensure it's okay regardless of how many threads their program has and how many times it's changing its own environ.

**Comment by tarcieri** [comment]
> I agree that it's misleading. It's vague, doesn't mention the conditions for the issue to occur, and links to an issue in the time crate's repo. Even now people are being confused and thinking the problem can be resolved by disabling the oldtime feature or updating the time dep to v0.3 ...

The advisory was filed in haste. If you would like to see it improved, the best way to do that is submitting PRs to the advisory.

**Comment by Dushistov** [comment]
> Request access to std::sys::unix::os::ENV_LOCK from e.g. the libs team, hoping some way of synchronizing access is added to future versions of std

How public `ENV_LOCK` can help? Crate can call `libc::set|get env` directly. Also C/C++ library can be linked with Rust crates into final executable, and `ENV_LOCK` will not be used by C/C++ libraries.

**Comment by tarcieri** [comment]
> How public ENV_LOCK can help? Crate can call libc::set|get env directly. 

The problem is crates are already making FFI calls to these APIs directly, or APIs that invoke them vicariously, without holding `ENV_LOCK`.

If they were holding `ENV_LOCK`, the access would be synchronized.

**Comment by briansmith** [comment]
> I found that `getenv` and `setenv` in libc are not thread-safe [1], and most impl of `localtime_r` in libc directly call `getenv` [2]. This means that `localtime_r` may have data race with `setenv`.

I strongly recommend reading the discussion in rust-lang/rust#27970. We seem to have converged on a consensus that reading the environment is safe because writing to the environment is unsafe and pretty much always wrong to do. IMO this was/is a non-issue and the real problem is in anything that calls env::set_var, and that env::set_var isn't marked unsafe. Again, that's explained in the 27970 issue.

**Comment by kolbma** [comment]
> > I found that `getenv` and `setenv` in libc are not thread-safe [1], and most impl of `localtime_r` in libc directly call `getenv` [2]. This means that `localtime_r` may have data race with `setenv`.
> 
> I strongly recommend reading the discussion in [rust-lang/rust#27970](https://github.com/rust-lang/rust/issues/27970). We seem to have converged on a consensus that reading the environment is safe because writing to the environment is unsafe and pretty much always wrong to do. IMO this was/is a non-issue and the real problem is in anything that calls env::set_var, and that env::set_var isn't marked unsafe. Again, that's explained in the 27970 issue.

And why it segfaults then?  
It shouldn't matter if there is something set or not set in the env in right order or time, there shouldn't be a SEGV!

**Comment by tarcieri** [comment]
Semi-related: on certain OSes, `chrono` calls `set_var`: https://github.com/chronotope/chrono/blob/3467172c31188006147585f6ed3727629d642fed/src/sys/unix.rs#L55-L78

That could probably use its own separate tracking issue. That usage pattern is not thread safe or memory safe.

**Comment by xfix** [comment]
> Semi-related: on certain OSes, `chrono` calls `set_var`:
> 
> https://github.com/chronotope/chrono/blob/3467172c31188006147585f6ed3727629d642fed/src/sys/unix.rs#L55-L78
> 
> That could probably use its own separate tracking issue. That usage pattern is not thread safe or memory safe.

Yeah, this races with itself even. The same code is in `time 0.1` crate too.

**Comment by tarcieri** [comment]
@Kixunil FWIW there has been quite a significant bit of discussion about various potential solutions on this thread:

https://internals.rust-lang.org/t/synchronized-ffi-access-to-posix-environment-variable-functions/15475

You might want to post your solution there for visibility, although I believe it might have already been discussed/attempted.

**Comment by Kixunil** [comment]
LOL, yes, I already did. :D

**Comment by briansmith** [comment]
I just filed #648, about the use of `env::set_var`, because I think *that* issue much more clearly needs to be addressed than this one.

For this issue (#499), we have to make a decision: Is it worth trying to make things "more often memory safe" in the case where the environment was being changed by Rust code, even though the general issue isn't fully addressed? Or are we going to say that changing the environment (in another thread) is just plain wrong and thus we're not going to do anything to make it work more often?

I would be fine with doing something like @Kixunil suggested, and I would also be OK with just closing this issue (#499) as "not a bug, working as intended, the bug is elsewhere."

**Comment by tarcieri** [comment]
The ideal fix to this particular issue would be a pure Rust reimplementation of `localtime_r`, and ideally one which can be shared (or provided by) the `time` crate.

That would ensure any environment mutation which occurs elsewhere via `env::set_var` is synchronized.

That said, it seems there is rough consensus that `env::set_var` should be deprecated in its current form and replaced with an `unsafe` alternative, which if it were to happen should hopefully encourage people not to use `env::set_var` capriciously. If that were to happen, it would help obviate the concern here.

**Comment by Kixunil** [comment]
> The ideal fix to this particular issue would be a pure Rust reimplementation of `localtime_r`

Yes and I think the next best is a reimplementation that happens to have some bits in C. It behaves the same and the only difference is we don't have a type proof of soundness.

I'm not convinced `set_var` should be `unsafe`/deprecated. If all software was written in Rust it'd be non-issue which indicates it's not a Rust design problem but a C problem. Or `unsafe` code problem if you look at it differently. All the rules around `unsafe` are meant to identify which piece of code is responsible for a memory problem. Safe code can never be responsible, thus the problem is with the code behind `unsafe` - one deciding to call into an unsound C library.

Deprecating `set_var` undermines the whole idea of safe-unsafe contracts. Or maybe the fact we have a discussion about it at all undermines it already.

**Comment by mathstuf** [comment]
> If all software was written in Rust it'd be non-issue which indicates it's not a Rust design problem but a C problem.

While we're positing hypotheticals, I think that if all software were written in Rust, I don't think we'd have had `set_var` in the first place. So I don't think that holds much water as to the state of `set_var`'s lack of an `unsafe` label.

> one deciding to call into an unsound C library.

And if the environment becomes Rust-only, Rust loses some of its power as a systems language because it cannot influence any process which is launched from non-Rust code anymore. Bad as the environment (and signals for that matter) are in POSIX land, they're a fact of life that one just has to live with today.

**Comment by tarcieri** [comment]
Here’s a pure Rust reimplementation of localtime: https://github.com/x-hgg-x/tz-rs

**Comment by djc** [comment]
I submitted #677 to address this issue.
